packages/component/src/provider.ts

```diff
@@ -0,0 +1,179 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Deferred } from '@firebase/util';
+import { ComponentContainer } from './component_container';
+import { DEFAULT_ENTRY_NAME } from './constants';
+import { InstantiationMode, Name, NameServiceMapping } from './types';
+import { Component } from './component';
+
+/**
+ * Provider for instance for service name T, e.g. 'auth', 'auth-internal'
+ * U is an alias for the type of the instance
+ */
+export class Provider<T extends Name = Name, U = NameServiceMapping[T]> {
```

Comment:
I think the existence of `U` here might confuse the compiler, making it think that U can actually be set to something else than `NameServiceMapping[T]`. Since we'd never want something like `Provider<'app', FirebaseInstallations>` I think we should remove `U` and replace its uses with `NameServiceMapping[T]`, even though it makes the code below more verbose.

Also I'd rather remove the default value for T, and type this as `Provider<Name>` explicitly where needed, which I  assume isn't that many places.

Reply:
Yeah, removed the generic. I wish there is a way to alias types within a class, something like:
```
class Provider<T extends Name = Name> {
   define type U = NameServiceMapping[T];
   // then use U as an alias in methods and properties
}
``` 🤷‍♂